src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -2007,31 +1970,22 @@ public void onLostInputs(LostInputsEvent event) {
     }
   }
 
-  /**
-   * Shuts the service down. Wait for active network I/O to finish but new requests are rejected.
-   */
+  /** Shuts the service down. */
   public void shutdown() {
     if (!shutdown.compareAndSet(false, true)) {
       return;
     }
 
     if (buildInterrupted.get()) {
+      backgroundTaskExecutor.shutdownNow();
       Thread.currentThread().interrupt();
     }
 
-    if (combinedCache != null) {
-      try {
-        backgroundTaskPhaser.awaitAdvanceInterruptibly(backgroundTaskPhaser.arrive());
-      } catch (InterruptedException e) {
-        buildInterrupted.set(true);
-        combinedCache.shutdownNow();
-        Thread.currentThread().interrupt();
-      }
+    // Waits for all background tasks to finish and interrupts them if there is another interrupt.
+    backgroundTaskExecutor.close();
 
-      // Only release the combinedCache once all background tasks have been finished. Otherwise, the
-      // last task might try to close the combinedCache inside the callback of network response
-      // which might cause deadlocks.
-      // See https://github.com/bazelbuild/bazel/issues/21568.
+    // Release the cache only after background tasks are done as they might be using it.
+    if (combinedCache != null) {
       combinedCache.release();
```

Comment:
Check interruption bit and call `combinedCache.shutdownNow` if the current thread is interrupted.